### ISSUE TITLE
ICE when running on cargo 

### ISSUE THREAD

**Issue by phansch** [rust-lang-nursery/rust-clippy#2774]
Split from #2603 to fix the ICE when running clippy on cargo.

An example is the integration test on Travis: https://travis-ci.org/rust-lang-nursery/rust-clippy/jobs/380402492#L1578

This was also reported here: https://github.com/rust-lang/rust/issues/50548

<details>
<summary>Click to expand :icecream: backtrace</summary>

```
error: internal compiler error: librustc/infer/region_constraints/mod.rs:649: cannot relate bound region: ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(1:1013), 'a)) <= ReScope(Remainder(BlockRemainder { block: ItemLocalId(937), first_statement_index: 0 }))
thread 'main' panicked at 'Box<Any>', librustc_errors/lib.rs:499:9
stack backtrace:
   0:     0x7f09b4111bfe - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h093c819cb048dde9
                               at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x7f09b40dc226 - std::sys_common::backtrace::print::h1677a20755935350
                               at libstd/sys_common/backtrace.rs:71
                               at libstd/sys_common/backtrace.rs:59
   2:     0x7f09b40e1dcb - std::panicking::default_hook::{{closure}}::h98084d605b12b847
                               at libstd/panicking.rs:211
   3:     0x7f09b40e1af6 - std::panicking::default_hook::h4d1f28d5902651c7
                               at libstd/panicking.rs:227
   4:     0x7f09b5c823c5 - rustc::util::common::panic_hook::h3ced0dc1f133801b
   5:     0x7f09b40e2505 - std::panicking::rust_panic_with_hook::h869eb4f5e1314587
                               at libstd/panicking.rs:467
   6:     0x7f09b5e8d90e - std::panicking::begin_panic::hcfe77e87ada22f89
   7:     0x7f09b57cad4a - rustc_errors::Handler::span_bug::hc3c9fe9a902533db
   8:     0x7f09b59cffb6 - rustc::session::opt_span_bug_fmt::{{closure}}::h9f90d13f5b9ed831
   9:     0x7f09b5a86f04 - rustc::ty::context::tls::with_opt::{{closure}}::haa004527159f9882
  10:     0x7f09b59d1def - rustc::ty::context::tls::with_context_opt::habfece4b87390b20
  11:     0x7f09b5a864e6 - rustc::ty::context::tls::with_opt::h112e7e73e5db6910
  12:     0x7f09b58f1ec4 - rustc::session::opt_span_bug_fmt::h41fbfa4394ebfc49
  13:     0x7f09b58f1e6f - rustc::session::span_bug_fmt::hbd84e0adf50782dd
  14:     0x7f09b5cf06a7 - rustc::infer::region_constraints::RegionConstraintCollector::make_subregion::hbcbea923f326e292
  15:     0x7f09b5cefc7e - rustc::infer::region_constraints::RegionConstraintCollector::make_eqregion::h2000dc855ddd74bb
  16:     0x7f09b58e9930 - <rustc::infer::equate::Equate<'combine, 'infcx, 'gcx, 'tcx> as rustc::ty::relate::TypeRelation<'infcx, 'gcx, 'tcx>>::regions::h78654076d77befdd
  17:     0x7f09b5971724 - rustc::ty::relate::super_relate_tys::hb599c498e7a41092
  18:     0x7f09b5c10722 - rustc::infer::combine::<impl rustc::infer::InferCtxt<'infcx, 'gcx, 'tcx>>::super_combine_tys::hf777672727656626
  19:     0x7f09b58e95ec - <rustc::infer::equate::Equate<'combine, 'infcx, 'gcx, 'tcx> as rustc::ty::relate::TypeRelation<'infcx, 'gcx, 'tcx>>::tys::h4068a583a1be54f1
  20:     0x7f09b582c79b - <rustc::ty::subst::Kind<'tcx> as rustc::ty::relate::Relate<'tcx>>::relate::heea06f7dd7b8a307
  21:     0x7f09b5cd8cd2 - <&'a mut I as core::iter::iterator::Iterator>::next::hf460ed909030202f
  22:     0x7f09b5d0b127 - <I as rustc::ty::context::InternAs<[T], R>>::intern_with::hbb4ef2f92c598e24
  23:     0x7f09b597165e - rustc::ty::relate::super_relate_tys::hb599c498e7a41092
  24:     0x7f09b5c10722 - rustc::infer::combine::<impl rustc::infer::InferCtxt<'infcx, 'gcx, 'tcx>>::super_combine_tys::hf777672727656626
  25:     0x7f09b58e95ec - <rustc::infer::equate::Equate<'combine, 'infcx, 'gcx, 'tcx> as rustc::ty::relate::TypeRelation<'infcx, 'gcx, 'tcx>>::tys::h4068a583a1be54f1
  26:     0x556ad3d29e06 - rustc::infer::InferCtxt::commit_if_ok::hf5f7530f87744cbc
  27:     0x556ad3d2a243 - rustc::infer::InferCtxt::can_eq::h1e863c0f3f7d3937
  28:     0x556ad3db6011 - rustc::ty::context::tls::with_related_context::h85c179df1cc83776
  29:     0x556ad3d29bf2 - rustc::infer::InferCtxtBuilder::enter::h25c5d396ec522472
  30:     0x556ad3df12a8 - clippy_lints::utils::same_tys::h3bd6ce3858318250
  31:     0x556ad3dc6d07 - <clippy_lints::types::ImplicitHasherConstructorVisitor<'a, 'b, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h9aeda3814f3faa82
  32:     0x556ad3d6583c - rustc::hir::intravisit::walk_decl::hc88aecb57bb15b1f
  33:     0x556ad3d7414a - rustc::hir::intravisit::walk_expr::hda984170f1b633eb
  34:     0x556ad3dc6cd0 - <clippy_lints::types::ImplicitHasherConstructorVisitor<'a, 'b, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h9aeda3814f3faa82
  35:     0x556ad3dc5c98 - <clippy_lints::types::ImplicitHasher as rustc::lint::LateLintPass<'a, 'tcx>>::check_item::hd767e19552856b4d
```

</details>

**Comment by oli-obk** [comment]
Idea: erase regions in the same_tys function. We don't care about lifetimes for type equality, do we?

**Comment by phansch** [comment]
Good question. I will try `erase_late_bound_regions` later today :+1: 

I already figured out that it's caused by the `HashSet` in [cargo/ops/resolve.rs:153](https://github.com/phansch/cargo/blob/master/src/cargo/ops/resolve.rs#L153) and really need to extract a minimal repro to debug this faster :/

**Comment by phansch** [comment]
I won't have time to work on this until Tuesday. If someone wants to pick this up, feel free! I have a WIP branch here: https://github.com/rust-lang-nursery/rust-clippy/compare/master...phansch:cannot_relate_bound_region_without_ICE_cream
